src/ripple/basics/LocalValue.h

```diff
@@ -0,0 +1,108 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_LOCALVALUE_H_INCLUDED
+#define RIPPLE_BASICS_LOCALVALUE_H_INCLUDED
+
+#include <boost/optional.hpp>
+#include <boost/thread/tss.hpp>
+#include <memory>
+#include <unordered_map>
+
+namespace ripple {
+
+namespace detail {
+
+struct LocalValues
+{
+    bool onCoro = true;
+
+    struct BasicValue
+    {
+        virtual ~BasicValue() = default;
+        virtual void* get() = 0;
+    };
+
+    template <class T>
+    struct Value : BasicValue
+    {
+        boost::optional<T> t;
+
+        void* get() override
+        {
+            return &t;
+        }
+    };
+
+    std::unordered_map<void const*, std::unique_ptr<BasicValue>> values;
+
+    static
+    inline
+    void
+    cleanup(LocalValues* lvs)
+    {
+        if (lvs && ! lvs->onCoro)
+            delete lvs;
+    }
+};
+
+template<class = void>
```

Comment:
`inline` tells the linker that multiple definitions are not an error. The compiler is free to use the the inline optimization technique anywhere it wants, and it is free to not use the inline optimization technique on functions declared `inline`. I really didn't want to start an argument here, but don't treat me like an idiot.